### PR TITLE
My solution of the GCDandLCM exercise

### PR DESCRIPTION
**Pull request by Maksim-Yeager**
If there is a problem, it might be the code documentation.

### CODE REVIEW

File: introducing-java/src/main/java/com/onto/javacourse/intro/gcd/GCD.java [hunk 1]
```diff
@@ -1,5 +1,33 @@
-public class Main {
-    public static void main(String[] args) {
+public class GCD {
```
Comment (ivorusev) [inline introducing-java/src/main/java/com/onto/javacourse/intro/gcd/GCD.java:9]: The class is missing a javadoc.

File: introducing-java/src/main/java/com/onto/javacourse/intro/gcd/GCD.java [hunk 2]
```diff
@@ -1,5 +1,33 @@
...
 
+    /**
+     * Method to find GCD
```
Comment (ivorusev) [inline introducing-java/src/main/java/com/onto/javacourse/intro/gcd/GCD.java]: The comment violates the java conventions. Method names start with a verb and end with a dot. I don't need to know from the comment that it is a method so this is redundant. For instance good comment would be `Calculates the greatest common divisor of two numbers.`.

File: introducing-java/src/main/java/com/onto/javacourse/intro/gcd/GCD.java [hunk 3]
```diff
@@ -1,5 +1,33 @@
...
+            while (a != b) {
+                if (a > b) {
+                    a -= b;
+                } else {
+                    b -= a;
+                }
+            }
+            return a;
+        } else {
+            throw new IllegalArgumentException("Invalid numbers");
```
Comment (ivorusev) [inline introducing-java/src/main/java/com/onto/javacourse/intro/gcd/GCD.java]: This is good, but I would swap the if the other way around. This way the code will be less nested in brackets.

Comment (ivorusev) [inline introducing-java/src/main/java/com/onto/javacourse/intro/gcd/GCD.java]: Why is the number invalid? Shouldn't the exception message provide some more information?

File: introducing-java/src/main/java/com/onto/javacourse/intro/gcd/GCD.java (hunk 2 above)
Comment (ivorusev) [reply introducing-java/src/main/java/com/onto/javacourse/intro/gcd/GCD.java]: You could add to the comment that you are using Euclidean algorithm for instance.

File: introducing-java/src/main/java/com/onto/javacourse/intro/gcd/GCD.java [hunk 4]
```diff
@@ -1,5 +1,33 @@
-public class Main {
-    public static void main(String[] args) {
+public class GCD {
 
+    /**
+     * Method to find GCD
+     */
+    public static int GCD(int a, int b) {
```
Comment (ivorusev) [inline introducing-java/src/main/java/com/onto/javacourse/intro/gcd/GCD.java]: a and b are bad for parameter names. Better would be to have something like firstNumber, secondNumber

File: introducing-java/src/main/java/com/onto/javacourse/intro/gcd/GCD.java [hunk 5]
```diff
@@ -1,5 +1,33 @@
...
+                }
+            }
+            return a;
+        } else {
+            throw new IllegalArgumentException("Invalid numbers");
+        }
     }
+
+    /**
+     * Method to find LCM
```
Comment (ivorusev) [inline introducing-java/src/main/java/com/onto/javacourse/intro/gcd/GCD.java]: Same issue with the javadoc as the one above.

File: introducing-java/src/main/java/com/onto/javacourse/intro/gcd/GCD.java [hunk 6]
```diff
@@ -1,5 +1,33 @@
...
+     */
+    public static int LCM(int a, int b) {
```
Comment (ivorusev) [inline introducing-java/src/main/java/com/onto/javacourse/intro/gcd/GCD.java]: Method name violates the naming conventions. Do you know why?

File: introducing-java/src/main/java/com/onto/javacourse/intro/gcd/GCD.java [hunk 7]
```diff
@@ -1,5 +1,33 @@
...
+        if (a != 0 || b != 0) {
+            return (a * b) / GCD(a, b);
+        } else {
+            throw new IllegalArgumentException("Invalid numbers");
```
Comment (ivorusev) [inline introducing-java/src/main/java/com/onto/javacourse/intro/gcd/GCD.java]: Again I'd swap the if.

File: introducing-java/src/test/java/com/onto/javacourse/intro/gcd/GCDTest.java [hunk 8]
```diff
@@ -0,0 +1,20 @@
+import org.junit.jupiter.api.Test;
+import static org.junit.jupiter.api.Assertions.*;
+
+public class GCDTest {
```
Comment (ivorusev) [inline introducing-java/src/test/java/com/onto/javacourse/intro/gcd/GCDTest.java:5]: This is good but you have an if statement in both methods. Can you also test the exception throwing, see https://www.baeldung.com/junit-assert-exception

**Review by ivorusev — CHANGES_REQUESTED** [review]

File: introducing-java/src/main/java/com/onto/javacourse/intro/gcd/GCD.java (hunk 6 above)
Comment (Maksim-Yeager) [reply introducing-java/src/main/java/com/onto/javacourse/intro/gcd/GCD.java]: Method names should start with a lowercase letter

File: introducing-java/src/main/java/com/onto/javacourse/intro/gcd/GCD.java [hunk 9]
```diff
@@ -1,5 +1,42 @@
...
+     *
+     * @param firstNumber  The first integer.
+     * @param secondNumber The second integer.
+     * @return The greatest common divisor of the two numbers.
+     * @throws IllegalArgumentException If both numbers are 0 or only one of them is 0.
+     */
+    public static int calculateGcd(int firstNumber, int secondNumber) {
+        if (firstNumber == 0 || secondNumber == 0) {
+            throw new IllegalArgumentException("Invalid numbers, they cannot be 0.");
+        } else {
```
Comment (ivorusev) [inline introducing-java/src/main/java/com/onto/javacourse/intro/gcd/GCD.java]: My idea was to remove the else clause, you can do that when throwing an exception

```
        if (firstNumber == 0 || secondNumber == 0) {
            throw new IllegalArgumentException("Invalid numbers, they cannot be 0.");
        }
        while ....
```

File: introducing-java/src/main/java/com/onto/javacourse/intro/gcd/GCD.java [hunk 10]
```diff
@@ -1,5 +1,42 @@
...
+     *
+     * @param firstNumber  The first integer.
+     * @param secondNumber The second integer.
+     * @return The least common multiple of the two numbers.
+     * @throws IllegalArgumentException If both numbers are 0 or only one of them is 0.
+     */
+    public static int calculateLcm(int firstNumber, int secondNumber) {
+        if (firstNumber == 0 || secondNumber == 0) {
+            throw new IllegalArgumentException("Invalid numbers, they cannot be 0.");
+        } else {
```
Comment (ivorusev) [inline introducing-java/src/main/java/com/onto/javacourse/intro/gcd/GCD.java]: Same as above, the else can be removed.

File: introducing-java/src/main/java/com/onto/javacourse/intro/gcd/GCD.java (hunk 1 above)
Comment (ivorusev) [reply introducing-java/src/main/java/com/onto/javacourse/intro/gcd/GCD.java:9]: Still missing a javaoc here. Classes should have one as well, the same was as methods

File: introducing-java/src/test/java/com/onto/javacourse/intro/gcd/GCDTest.java [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+
+        String expectedMessage = "Invalid numbers, they cannot be 0.";
+        String actualMessageBothZero = exceptionBothZero.getMessage();
+        String actualMessageFirstIsZero = exceptionFirstIsZero.getMessage();
+        String actualMessageSecondIsZero = exceptionSecondIsZero.getMessage();
+
+        assertTrue(actualMessageBothZero.contains(expectedMessage));
+        assertTrue(actualMessageFirstIsZero.contains(expectedMessage));
+        assertTrue(actualMessageSecondIsZero.contains(expectedMessage));
+
```
Comment (ivorusev) [inline introducing-java/src/test/java/com/onto/javacourse/intro/gcd/GCDTest.java:32]: You don't need this new line. New lines are typically used to split logical parts in the code and this one is redundant.

**Review by ivorusev — CHANGES_REQUESTED** [review]

**Review by ivorusev — APPROVED** [review]

**Pull request merged**